hpc0.qmd

```diff
@@ -7,279 +7,1186 @@ format:
+```
+- git is a distributed version control system
+- Change directory into the directory called `shell-training`.
+- List the directory contents.
+
+
+## `cat`, `sort` and `cut` {.smaller}
+- The `cat` command will concatenate and print to standard output.
+  - This can be used to stream one thing after another.
+- The `sort` command will sort lines (by default alphabetically).
+- The `cut` command will cut lines based on a given character.
+- Commands that create an output stream can have the output piped into another command that can process the stream using the symbol '`|`'. 
+  - For example two files `test` and `test2` can be concatenated, sorted alphabetically and piped into head as follows: 
+```{.bash code-line-numbers=false}
+cat test test2 | sort | head
+```
+
+
+## `cat`, `sort` and `cut` demo {visibility="hidden"}
+::: {.notes}
+- Offer a demo?
+:::
+
+
+## `cat`, `sort` and `cut` exercise {.smaller}
 
 ::: {.panel-tabset}
 
 ### Problem
+- Combine `cat`, `cut`, and `sort` to print out the Latin names from the file `IOM-animals/birds.txt` in alphabetical order.
+- Save the output to a new file.
+Hint: `cut -d ',' -f 2` will split lines around the comma and yield the second field.
 
-`shell-training/data/` contains 300 data files, each of which should contain 100 values. One of these files is missing some data though. . .
-
+### Solution
+```{.bash code-line-numbers=false}
+cat birds.txt | sort | cut -d ',' -f 2 > sorted-birds.txt
+```
+\
 \
-Use a series of commands connected by pipes to identify the file with missing data. Hints:
+- Another option to achieve the same is to use `awk` instead of `cut`:
+```{.bash code-line-numbers=false}
+cat birds.txt | sort | awk '{FS=","}{print $2}' > sorted-birds.txt
+```
+- `awk` is a scripting language used for manipulating data and generating reports.
+  - It supports variables, numeric functions, string functions, and logical operators.
+  - It allows for setting field seperators of multiple characters, whereas `cut` is restricted to a single character field seperator. 
+- The '`{FS=","}`' part of the `awk` command sets the field seperator to a comma, to set as a space followed by a comma use '`{FS=" ,"}`'
 
-- `wc -w` will tell you the number of values in a file
-- `sort -n` will sort numerically
+:::
+
+
+## `head` and `tail` {.smaller}
+- The `head` command prints the first lines of a file.
+- The `tail` command prints the last lines of a file.
+- The default number of lines is 10.
+ - This can be changed with the '`-n`' option
+
+
+## `head` and `tail` exercise {.smaller}
+
+::: {.panel-tabset}
+
+### Problem
+- List all the lines in the files in `IOM-animals` directory alphabetically and find the `50th` item.
 
 ### Solution
+- Use `head` to get the first 50 lines and `pipe it` to `tail` to get the last one:
+```{.bash code-line-numbers=false}
+sort *.txt | head -n 50 | tail -n 1
+```
+\
+\
+- Another option is to use the `sed` command instead of `head` and `tail`:
```

Comment:
We get the same bullet point oddness we have before.  Just make this a string and not a bullet point?